snapshot-service/src/lib.rs

```diff
@@ -0,0 +1,284 @@
+        let entries = db.open_tree("entries")?;
+
+        let mut next_tag_id = None;
+
+        for kv in tags.iter() {
+            let (_, v) = kv?;
+
+            let id = u32::from_be_bytes(v.as_ref().try_into().unwrap());
+
+            next_tag_id = next_tag_id.map(|c| max(c, id)).or_else(|| Some(id));
+        }
+
+        let next_tag_id = next_tag_id.map(|id| id + 1).unwrap_or(0);
+
+        Ok(UpdateHandle {
+            _db: db,
+            tags,
+            entries,
+            next_tag_id,
+        })
+    }
+
+    #[tracing::instrument(skip(self, snapshot))]
+    pub async fn update(
+        &mut self,
+        tag: &str,
+        snapshot: impl IntoIterator<Item = VoterHIR>,
+    ) -> Result<(), Error> {
+        let mut batch = sled::Batch::default();
+
+        let tag_id = match self.tags.get(tag.clone())? {
+            Some(tag_id_raw) => {
+                let tag_id = u32::from_be_bytes(tag_id_raw.as_ref().try_into().unwrap());
+
+                for entry in self
+                    .entries
+                    .range(tag_id_raw.as_ref()..&(tag_id + 1).to_be_bytes())
+                {
+                    let (k, _) = entry?;
+
+                    batch.remove(k);
+                }
+
+                Some(tag_id_raw)
+            }
+            None => None,
+        };
+
+        for entry in snapshot.into_iter() {
+            let VoterHIR {
+                voting_key,
+                voting_group,
+                voting_power,
+            } = entry;
+
+            let mut key = Vec::with_capacity(4 + 32 + voting_group.as_bytes().len());
+
+            match tag_id {
+                None => key.extend(&self.next_tag_id.to_be_bytes()),
+                Some(ref tag_id_raw) => key.extend(&**tag_id_raw),
+            }
+
+            key.extend(voting_key.as_ref().as_ref());
+            key.extend(voting_group.as_bytes());
+
+            batch.insert(key, &u64::from(voting_power).to_be_bytes());
+        }
+
+        {
+            let tag = tag.to_string();
+            let tag_id = tag_id.clone();
+            let tags = self.tags.clone();
+            let entries = self.entries.clone();
+            let next_tag_id = self.next_tag_id.to_be_bytes();
+
+            tokio::task::spawn_blocking(move || {
+                (&tags, &entries).transaction(move |(tags, entries)| {
+                    if let None = tag_id {
+                        tags.insert(tag.as_bytes(), &next_tag_id)?;
+                    }
+
+                    entries.apply_batch(&batch)?;
+                    Ok(())
+                })?;
+
+                Ok(())
+            })
+            .await
+            .unwrap()
+            .map_err(Error::DbTxError)?;
+        }
+
+        if let None = tag_id {
+            self.next_tag_id += 1;
+        }
+
+        Ok(())
+    }
+}
+
+pub fn new_context() -> Result<(SharedContext, UpdateHandle), Error> {
+    let db = sled::Config::new().temporary(true).open()?;
```

Comment:
could a disk-backed db be useful if we have to restart the service?

Reply:
it is disk-backed even with temporary, if it was in memory I would have used an normal std Hash/BTree map as I was using a few commits ago. 

And yes, but with the current implementation it will read all the files in the snapshots directory at bootstrap, so it's not really much of a difference except for saving some time. If we switch to an endpoint to update then it would be more useful.

Reply:
Correct, I was indeed thinking of the second option, but at least we've found some advantage of using the file based trigger

Reply:
But IIRC it should be quite easy to add persistence to this, right?